shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java

```diff
@@ -404,91 +415,247 @@ public void run() {
     }
   }
 
+  // Keep a queue of ImageReaders.
+  // Each ImageReader holds acquired Images.
+  // When we acquire the next image, close any ImageReaders that don't have any
+  // more pending images.
   @Keep
   @TargetApi(29)
   final class ImageReaderSurfaceProducer
       implements TextureRegistry.SurfaceProducer, TextureRegistry.ImageConsumer {
     private static final String TAG = "ImageReaderSurfaceProducer";
     private static final int MAX_IMAGES = 4;
 
+    private static final boolean VERBOSE_LOGS = false;
+
     private final long id;
 
     private boolean released;
     private boolean ignoringFence = false;
 
     private int requestedWidth = 0;
     private int requestedHeight = 0;
+    private boolean createNewReader = true;
+
+    private long lastDequeueTime = 0;
+    private long lastQueueTime = 0;
+    private long lastScheduleTime = 0;
+
+    private Object lock = new Object();
+    // REQUIRED: The following fields must only be accessed when lock is held.
+    private final LinkedList<PerImageReader> imageReaderQueue = new LinkedList<PerImageReader>();
+    private final HashMap<ImageReader, PerImageReader> perImageReaders =
+        new HashMap<ImageReader, PerImageReader>();
 
-    /** Internal class: state held per image produced by image readers. */
+    /** Internal class: state held per Image produced by ImageReaders. */
     private class PerImage {
-      public final ImageReader reader;
       public final Image image;
+      public final long queuedTime;
 
-      public PerImage(ImageReader reader, Image image) {
-        this.reader = reader;
+      public PerImage(Image image, long queuedTime) {
         this.image = image;
+        this.queuedTime = queuedTime;
+      }
+    }
+
+    /** Internal class: state held per ImageReader. */
+    private class PerImageReader {
```

Comment:
This is a complex enough Java class where I'd expect a unit test suite for it.

Maybe we have enough test coverage in this PR, but it's not clear to me.

Reply:
Ack.